File: Languages/cpp/algorithms-dsa/leetcode/greedy/jump_game.cpp
/**
 * LeetCode 55: Jump Game
 * Topic: Greedy
 * Difficulty: Medium
 *
 * Given an array where each element is the max jump length from that position,
 * determine if you can reach the last index. Track farthest reachable position.
 * Time: O(n), Space: O(1)
 */
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

bool canJump(vector<int>& nums) {
    int farthest = 0;
    for (int i = 0; i < (int)nums.size(); i++) {
        if (i > farthest) return false;
        farthest = max(farthest, i + nums[i]);
    }
    return true;
}

int main() {
    vector<int> v1 = {2, 3, 1, 1, 4};
    assert(canJump(v1) == true);

    vector<int> v2 = {3, 2, 1, 0, 4};
    assert(canJump(v2) == false);

    vector<int> v3 = {0};
    assert(canJump(v3) == true);

    vector<int> v4 = {2, 0, 0};
    assert(canJump(v4) == true);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/greedy/jump_game.py
"""
LeetCode #55 - Jump Game
Topic: Greedy
Difficulty: Medium

Determine if you can reach the last index by tracking farthest reachable position.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def can_jump(nums: list[int]) -> bool:
    farthest = 0
    for i in range(len(nums)):
        if i > farthest:
            return False
        farthest = max(farthest, i + nums[i])
    return True


if __name__ == "__main__":
    assert can_jump([2, 3, 1, 1, 4]) is True
    assert can_jump([3, 2, 1, 0, 4]) is False
    assert can_jump([0]) is True
    assert can_jump([2, 0, 0]) is True
    print("All tests passed!")
